velox/functions/prestosql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,325 @@
+
+    for (auto i = 0; i < actualVector.size(); i++) {
+      ASSERT_TRUE(reader.isSet(i))
+          << "Actual array at " << i
+          << " is a null, which we don't need to test here";
+      auto actualArray = reader[i].materialize();
+      // 3. assert the two vectors containing the same elements (ignoring order)
+      LOG(WARNING) << "Actual array " << printArray(actualArray) << " at " << i
+                   << " vs expected array " << printArray(expected[i]);
+      ASSERT_TRUE(std::is_permutation(
+          actualArray.begin(), actualArray.end(), expected[i].begin()))
+          << "Actual array " << printArray(actualArray) << " at " << i
+          << " must produce a permutation of the expected array "
+          << printArray(expected[i]);
+    }
+  }
+
+  template <typename T, typename Cont = Array<T>>
+  void testIsPermutation(
+      const std::vector<std::vector<std::optional<T>>>& expected,
+      const std::string& expression,
+      const std::vector<VectorPtr>& input) {
+    auto arrayVectorResult =
+        evaluate<ArrayVector>(expression, makeRowVector(input));
+    decodeElementsAndTestIsAPermutation<T, Cont>(
+        *arrayVectorResult.get(), expected);
+  }
+
+  /// To test a shuffle, we will do the following steps:
+  /// 1. run the test for T times with range-based input {0, 1, ..., n-1},
+  ///    and it should be at least X% times different in total.
+  ///    Let's say T = 100, n = 10, X = 70.
+  /// 2. test if each shuffle array is a permutation of the origin array,
+  ///    which ensures the shuffle will not miss any existing value or
+  ///    generate any unexpected values.
+  /// NOTE:
+  /// The combination of tests above is a naive but straightforward way of
+  /// verifying the shuffle's (1) randomness and (2) and correctness. However,
+  /// it doesn't guarantee the uniformity. We will avoid over testing with
+  /// leveraging standard deviation, etc.
+  void testRandomness(
+      const int32_t t = kNumOfEvalTimes,
+      const int32_t n = kNums) {
+    // generate a range-based array {0, 1, ..., n-1} and shuffle it
+    std::vector<int32_t> actual(n);
+    std::iota(actual.begin(), actual.end(), 0);
+    std::vector<int32_t> expected(actual);
+    std::mt19937 rng(std::random_device{}());
+    std::shuffle(actual.begin(), actual.end(), rng);
+
+    const int32_t kThreshold = (int32_t)(n * kUniquenessRate);
+    auto actualVector = makeArrayVector<int32_t>({actual});
+    folly::F14FastSet<std::vector<int32_t>> distinctResults;
+
+    for (auto i = 0; i < t; i++) {
+      auto arrayVectorResult =
+          evaluate<ArrayVector>("shuffle(C0)", makeRowVector({actualVector}));
+      // decode element values from output ArrayVector
+      DecodedVector decoded;
+      exec::VectorReader<Any> reader(decode(decoded, *arrayVectorResult));
+      auto arrayReader = reader[0].castTo<Array<int32_t>>();
+      std::vector<int32_t> shuffledResult(kNums);
```

Comment:
Any particular reason not to use VectorReader::materialize?

Reply:
Thanks for raising it. 
One gotcha: since my input and output arrays are both null-free, I found instead of calling `reader[0]`, I need to call `reader.readNullFree(0)` to force it return a `std::vector<int32_t>`.